File: benchmark/harness/Sources/BenchmarkKit/Generators.swift
// Generators.swift — deterministic WITH/WITHOUT generators, port of
// benchmark/with-plugin/generate.py + without-plugin/generate.py.
//
// WITH path: drives the REAL estimate-calc.py (subprocess) for the complexity
// score + estimated ai_cost, then materializes the ttt-template as N staged
// operations (stage_count = process-overhead signal, >1).
// WITHOUT path: single-shot copy, stage_count == 1, complexity score 0.
// Both copy benchmark/ttt-template/ (excluding .build/.swiftpm) into the workdir.
//
// Deterministic mode only. No network, no LLM, never imports BenchmarkLive.

import Foundation

public enum Generators {
    /// Five complexity factors for the TTT workload (small, well-understood).
    static let factors = ["3", "3", "3", "3", "3"]
    static let expectedTokens = "100000"

    /// OI-1 (P1): run `body` against a freshly-created ephemeral workdir and
    /// remove that workdir on EVERY exit path — normal return, thrown error, or
    /// early return inside `body`. Callers that generate a throwaway app copy
    /// (tests, ad-hoc measurements) use this instead of hand-rolling a
    /// `NSTemporaryDirectory()` dir they then leak (the two ENOSPC live kills).
    /// The `defer` is scoped to the workdir this helper OWNS — a persistent
    /// `--workdir` passed to `generateWith*` directly is never touched (R3).
    @discardableResult
    public static func withEphemeralWorkdir<T>(
        prefix: String = "ttt_gen_",
        _ body: (_ workdir: String) throws -> T
    ) throws -> T {
        let fm = FileManager.default
        let workdir = (NSTemporaryDirectory() as NSString)
            .appendingPathComponent(prefix + UUID().uuidString)
        try fm.createDirectory(atPath: workdir, withIntermediateDirectories: true)
        defer { try? fm.removeItemIfExists(atPath: workdir) }
        return try body(workdir)
    }

    /// Drive the REAL estimate-calc.py. Returns (complexityTotal, aiCostUSD?).
    /// On any failure the cost degrades to nil (never fabricated) — the score
    /// falls back to 0 so a broken estimate never crashes a deterministic run.
    public static func runEstimate(estimateCalcPath: String,
                                   runner: (([String]) -> String)? = nil)
        -> (score: Int, cost: Double?) {
        let argv = [
            "python3", estimateCalcPath,
            "--size", "M", "--level", "senior",
            "--factors", factors[0], factors[1], factors[2], factors[3], factors[4],
            "--tokens", expectedTokens, "--model", "sonnet",
        ]
        let stdout: String
        if let runner {
            stdout = runner(argv)
        } else {
            let r = Subprocess.run(argv)
            guard r.exitCode == 0 else { return (0, nil) }
            stdout = r.stdout
        }
        guard let parsed = try? JSONParser.parse(stdout) else { return (0, nil) }
        let score = parsed["complexity"]?["total"]?.intValue ?? 0
        let cost = parsed["ai_cost"]?["usd"]?.doubleValue
        return (score, cost)
    }

    /// WITH generator: staged scaffold driven by the real estimate.
    public static func generateWithPlugin(
        workdir: String, templateDir: String, pluginRoot: String, estimateCalcPath: String,
        estimateRunner: (([String]) -> String)? = nil
    ) throws -> PathMetrics {
        let appDir = (workdir as NSString).appendingPathComponent("with")
        let timer = GenLib.Timer(); timer.start()
        let (score, _) = runEstimate(estimateCalcPath: estimateCalcPath, runner: estimateRunner)
        let stageCount = try GenLib.copyTemplateStaged(templateDir: templateDir, dest: appDir)
        let elapsed = timer.elapsed
        // Deterministic mode records cost as null (estimate is not real spend).
        return GenLib.buildPathMetrics(
            appDir: appDir, pluginRoot: pluginRoot, stageCount: stageCount,
            estimateComplexityScore: score, costUSD: nil, wallClockS: elapsed
        )
    }

    /// WITHOUT generator: single-shot copy, no estimate, no staging.
    public static func generateWithoutPlugin(
        workdir: String, templateDir: String, pluginRoot: String
    ) throws -> PathMetrics {
        let appDir = (workdir as NSString).appendingPathComponent("without")
        let timer = GenLib.Timer(); timer.start()
        let stageCount = try GenLib.copyTemplateSingleShot(templateDir: templateDir, dest: appDir)
        let elapsed = timer.elapsed
        return GenLib.buildPathMetrics(
            appDir: appDir, pluginRoot: pluginRoot, stageCount: stageCount,
            estimateComplexityScore: 0, costUSD: nil, wallClockS: elapsed
        )
    }
}

File: benchmark/harness/benchmarkkit/generators.py
"""Deterministic WITH/WITHOUT generators.

WITH: drives the REAL estimate-calc.py (subprocess) for the complexity score, then
materializes ttt-template as N staged ops (stage_count >1). WITHOUT: single-shot
copy, stage_count == 1, score 0. Both copy benchmark/ttt-template/ into the workdir.
Deterministic only — no network, no LLM, never imports benchmarklive.
"""

from __future__ import annotations

import contextlib
import json
import os
import shutil
import tempfile
from typing import Callable, Optional

from . import genlib, treecopy
from .genlib import Subprocess, Timer
from .metrics import PathMetrics

FACTORS = ["3", "3", "3", "3", "3"]
EXPECTED_TOKENS = "100000"


@contextlib.contextmanager
def with_ephemeral_workdir(prefix: str = "ttt_gen_"):
    """Yield a freshly-created workdir removed on EVERY exit path (OI-1/R3).

    Scoped to the dir this helper owns; a persistent --workdir is never touched.
    """
    workdir = tempfile.mkdtemp(prefix=prefix)
    try:
        yield workdir
    finally:
        shutil.rmtree(workdir, ignore_errors=True)


def run_estimate(estimate_calc_path: str,
                 runner: Optional[Callable[[list], str]] = None) -> tuple:
    """Drive the REAL estimate-calc.py. Returns (score, cost_usd|None).

    Any failure degrades cost to None (never fabricated); score falls back to 0.
    """
    argv = [
        "python3", estimate_calc_path,
        "--size", "M", "--level", "senior",
        "--factors", *FACTORS,
        "--tokens", EXPECTED_TOKENS, "--model", "sonnet",
    ]
    if runner is not None:
        stdout = runner(argv)
    else:
        r = Subprocess.run(argv)
        if r.exit_code != 0:
            return 0, None
        stdout = r.stdout
    try:
        parsed = json.loads(stdout)
    except (ValueError, TypeError):
        return 0, None
    score = (parsed.get("complexity") or {}).get("total") or 0
    cost = (parsed.get("ai_cost") or {}).get("usd")
    return score, cost


def _grade(app_dir: str, grader: Optional[Callable]) -> Optional[dict]:
    """Score the generated app; ``None`` grader leaves the record's oracle absent."""
    return None if grader is None else grader(app_dir).to_dict()


def generate_with_plugin(workdir: str, template_dir: str, plugin_root: str,
                         estimate_calc_path: str,
                         estimate_runner: Optional[Callable[[list], str]] = None,
                         oracle_grader: Optional[Callable] = None) -> PathMetrics:
    app_dir = os.path.join(workdir, "with")
    timer = Timer()
    timer.start()
    score, _ = run_estimate(estimate_calc_path, runner=estimate_runner)
    stage_count = treecopy.copy_template_staged(template_dir, app_dir)
    elapsed = timer.elapsed
    # Deterministic mode records cost as null (estimate is not real spend).
    return genlib.build_path_metrics(
        app_dir=app_dir, plugin_root_dir=plugin_root, stage_count=stage_count,
        estimate_complexity_score=score, cost_usd=None, wall_clock_s=elapsed,
        oracle_result=_grade(app_dir, oracle_grader),
    )


def generate_without_plugin(workdir: str, template_dir: str, plugin_root: str,
                            oracle_grader: Optional[Callable] = None) -> PathMetrics:
    app_dir = os.path.join(workdir, "without")
    timer = Timer()
    timer.start()
    stage_count = treecopy.copy_template_single_shot(template_dir, app_dir)
    elapsed = timer.elapsed
    return genlib.build_path_metrics(
        app_dir=app_dir, plugin_root_dir=plugin_root, stage_count=stage_count,
        estimate_complexity_score=0, cost_usd=None, wall_clock_s=elapsed,
        oracle_result=_grade(app_dir, oracle_grader),
    )
